File: Source/Aladdin.CAPI.PKCS11/Provider.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis; 
using Aladdin.PKCS11; 

namespace Aladdin.CAPI.PKCS11
{
	///////////////////////////////////////////////////////////////////////////
	// Криптографический провайдер
	///////////////////////////////////////////////////////////////////////////
    [SuppressMessage("Microsoft.Design", "CA1063:ImplementIDisposableCorrectly")]
	public abstract class Provider : CryptoProvider, IRandFactory
	{
		// конструктор
		protected Provider(string name) { this.name = name; } private string name;

        // имя провайдера
        public override string Name { get { return name; }}
		// интерфейс вызова функций
		public abstract Module Module { get; } 

        // возможность импорта ключевой пары в память
        public virtual bool CanImportSessionPair(Applet applet) { return true; }

		///////////////////////////////////////////////////////////////////////////
		// Создание атрибутов
		///////////////////////////////////////////////////////////////////////////
		public Attribute CreateAttribute(ulong type, byte value)
		{
			// создать атрибут
			return new Attribute(type, value); 
		}
		public Attribute CreateAttribute(ulong type, byte[] value)
		{
			// создать атрибут
			return new Attribute(type, value); 
		}
        public Attribute CreateAttribute(ulong type, string value)
		{
			// создать атрибут
			return new Attribute(type, value); 
		}
		public Attribute CreateAttribute(ulong type, ulong value)
        {
	        // создать атрибут
	        return new Attribute(type, Module.EncodeLong(value)); 
        }
		///////////////////////////////////////////////////////////////////////
		// Атрибуты провайдера
		///////////////////////////////////////////////////////////////////////

		// фабрика генераторов случайных данных
		public override IRandFactory CreateRandFactory(SecurityObject scope, bool strong)
        {
	        // проверить область видимости
	        if (scope is Container) return RefObject.AddRef(((Container)scope).Store);
	        
	        // проверить область видимости
	        if (scope is Applet) return RefObject.AddRef((Applet)scope); 
            
            // вернуть фабрику генераторов
            return RefObject.AddRef(this);
        }
        // создать генератор случайных данных
        public override IRand CreateRand(object window)
        { 
	        // получить список считывателей
	        UInt64[] slotList = Module.GetSlotList(true); 
	
	        // для каждого считывателя
	        for (int i = 0; i < slotList.Length; i++)
	        try {
                // получить информацию устройства
                SlotInfo slotInfo = Module.GetSlotInfo(slotList[i]); 
            
                // проверить наличие смарт-карты
                if ((slotInfo.Flags & API.CKF_TOKEN_PRESENT) == 0) continue;

		        // получить информацию устройства
		        TokenInfo tokenInfo = Module.GetTokenInfo(slotList[i]); 

		        // проверить наличие генератора случайных данных
		        if ((tokenInfo.Flags & API.CKF_RNG) == 0) continue; 

                // создать объект смарт-карты
                using (Token token = new Token(this, slotList[i])) 
                {
		            // создать объект апплета
                    using (Applet applet = new Applet(token, slotList[i]))  
                    {
		                // вернуть генератор случайных данных
		                return applet.CreateRand(window);
                    }
                }
	        }
	        // вызвать базовую функцию
	        catch {} return base.CreateRand(window); 
        }
		///////////////////////////////////////////////////////////////////////
		// Управление объектами
		///////////////////////////////////////////////////////////////////////

		// получить хранилище контейнера
		public override string[] EnumerateStores(Scope scope)
        {
            // проверить область видимости
            if (scope != Scope.System) return new string[0]; 

            // выделить память для имен смарт-карт
            List<String> stores = new List<String>(); 
            try { 
	            // получить список считывателей
	            UInt64[] slotList = Module.GetSlotList(true); 

	            // для всех найденных смарт-карт
	            for (int i = 0; i < slotList.Length; i++) 
	            try {
		            // создать объект считывателя
		            using (Slot slot = new Slot(this, slotList[i]))
                    { 
		                // при наличии смарт-карты
		                if (slot.GetState() != PCSC.ReaderState.Card) continue; 
		
		                // добавить смарт-карту в список
		                if (!stores.Contains(slot.Name)) stores.Add(slot.Name); 
                    }
	            }
	            catch {} 
            } 
            // вернуть список смарт-карт
            catch {} return stores.ToArray(); 
        }
		// получить хранилище контейнера
		public override SecurityStore OpenStore(Scope scope, string storeName)
        {
	        // проверить область видимости
	        if (scope != Scope.System) throw new NotFoundException(); 

	        // получить список считывателей
	        UInt64[] slotList = Module.GetSlotList(true); 

	        // для всех найденных смарт-карт
	        for (int i = 0; i < slotList.Length; i++) 
	        {
		        // получить имя считывателя
		        SlotInfo info = Module.GetSlotInfo(slotList[i]); 

		        // проверить совпадение имен
		        if (!info.SlotDescription.Equals(storeName)) continue; 

		        // открыть объект смарт-карты
		        return new Token(this, slotList[i]); 
	        }
	        // при ошибке выбросить исключение
	        throw new NotFoundException(); 
        }
		// найти устройство с реализацией алгоритма
		public Applet FindApplet(SecurityObject scope, ulong algID, ulong usage, int keySize)
        {
	        // проверить область видимости
	        if (scope is Container) 
	        {
                // получить апплет контейнера
                Applet applet = ((Container)scope).Store; 

                // проверить поддержку алгоритма
                if (!applet.Supported(algID, usage, keySize)) return null; 

		        // вернуть устройство для контейнера
		        return RefObject.AddRef(applet);
	        }
	        // проверить область видимости
            if (scope is Applet) { Applet applet = (Applet)scope; 
            
                // проверить поддержку алгоритма
                if (!applet.Supported(algID, usage, keySize)) return null; 

		        // вернуть устройство 
		        return RefObject.AddRef(applet);
            }
	        // получить список считывателей
	        UInt64[] slotList = Module.GetSlotList(true); 
	
	        // для каждого считывателя
	        for (int i = 0; i < slotList.Length; i++)
	        try {
                // получить информацию устройства
                SlotInfo slotInfo = Module.GetSlotInfo(slotList[i]); 
            
                // проверить наличие смарт-карты
                if ((slotInfo.Flags & API.CKF_TOKEN_PRESENT) == 0) continue;
                 
                // создать объект смарт-карты
                using (Token token = new Token(this, slotList[i])) 
                {
                    // открыть объект апплета
                    using (Applet applet = new Applet(token, slotList[i]))
                    {
                        // проверить поддержку алгоритма
                        if (!applet.Supported(algID, usage, keySize)) continue; 
                        
                        // вернуть найденный апплет
                        return RefObject.AddRef(applet);
                    }
                }
            }
            catch {} return null;
        }
		///////////////////////////////////////////////////////////////////////
		// Особенности провайдера
		///////////////////////////////////////////////////////////////////////

		// сгенерировать стартовое значение
		public virtual byte[] GenerateSeed(Applet applet) { return null; } 

		// преобразование типа ключей
		public virtual SecretKey ConvertSecretKey(SessionObject obj, SecretKeyFactory keyFactory)
        {
	        // получить атрибуты ключа
	        Attributes attributes = GetKeyAttributes(obj, 
                CreateAttribute(API.CKA_KEY_TYPE, API.CKK_GENERIC_SECRET) 
            ); 
            // при возможности извлечения значения
            if (attributes[API.CKA_EXTRACTABLE].GetByte() != API.CK_FALSE && 
                attributes[API.CKA_SENSITIVE  ].GetByte() == API.CK_FALSE)
            {
                // получить значение ключа
                Attribute attribute = CreateAttribute(API.CKA_VALUE, obj.GetValue()); 
            
                // добавить атрибут в список
                attributes = attributes.Join(attribute); 
            }
            // при отсутствии на смарт-карте
            if (attributes[API.CKA_TOKEN].Value[0] == 0)
            {
                // проверить наличие значения
                if (attributes[API.CKA_VALUE] == null)
                {
                    // при ошибке выбросить исключение
                    throw new Aladdin.PKCS11.Exception(API.CKR_KEY_UNEXTRACTABLE); 
                }
		        // создать ключ по сеансовому объекту 
		        return new SecretKey(null, keyFactory, attributes); 
	        }
	        else {
                // определить идентификатор слота
                UInt64 slotID = obj.Session.SlotID; 

		        // открыть объект смарт-карты
		        using (Token token = new Token(this, slotID)) 
                {
                    // указать апплет
                    using (Applet applet = new Applet(token, slotID))  
                    {
		                // создать ключ по сеансовому объекту 
		                return new SecretKey(applet, keyFactory, attributes);
                    }
                }
	        }
        }
        // преобразовать тип открытого ключа
		public abstract IPublicKey ConvertPublicKey(Applet applet, SessionObject obj); 

        // преобразовать тип личного ключа
		public abstract PrivateKey ConvertPrivateKey(SecurityObject scope, 
            SessionObject privateObject, IPublicKey publicKey
        ); 
		// преобразование типа ключей
		public SessionObject ToSessionObject(
            Session session, ISecretKey key, Attribute[] keyAttributes)
        {
	        // проверить тип ключа
	        if (key is SecretKey) return ((SecretKey)key).ToSessionObject(session, keyAttributes); 

            // получить значение ключа
            if (key.Value == null) throw new Aladdin.PKCS11.Exception(API.CKR_KEY_UNEXTRACTABLE); 

            // создать список атрибутов
            Attribute[] attributes = new Attribute[] { 

	            // указать тип ключа
	            CreateAttribute(API.CKA_CLASS, API.CKO_SECRET_KEY), 

	            // указать тип ключа
	            CreateAttribute(API.CKA_KEY_TYPE, API.CKK_GENERIC_SECRET), 

	            // указать извлекаемость значения
	            CreateAttribute(API.CKA_EXTRACTABLE, API.CK_TRUE ), 
	            CreateAttribute(API.CKA_SENSITIVE  , API.CK_FALSE), 

                // указать значение ключа
	            CreateAttribute(API.CKA_VALUE, key.Value) 
            }; 
            // создать сеансовый объект
            return session.CreateObject(Attribute.Join(attributes, keyAttributes)); 
        }
		public SessionObject ToSessionObject(Session session, 
            IPublicKey publicKey, MechanismInfo info, Attribute[] keyAttributes)
        {
            // получить атрибуты открытого ключа
            Attribute[] publicKeyAttributes = PublicKeyAttributes(
                null, publicKey, info
            ); 
            // проверить поддержку ключа
            if (publicKeyAttributes == null) throw new NotSupportedException(); 

            // указать тип ключа
	        Attribute[] attributes = new Attribute[] {
		        CreateAttribute(API.CKA_CLASS, API.CKO_PUBLIC_KEY) 
	        }; 
            // добавить атрибуты открытого ключа
	        attributes = Attribute.Join(attributes, publicKeyAttributes); 

            // создать сеансовый объект
            return session.CreateObject(Attribute.Join(attributes, keyAttributes)); 
        }
		public SessionObject ToSessionObject(Session session, 
            IPrivateKey privateKey, MechanismInfo info, Attribute[] keyAttributes)
        {
	        // проверить тип ключа
            if (privateKey is PrivateKey) { PrivateKey pkcs11Key = (PrivateKey)privateKey; 
            
                // создать сеансовый объект
                return pkcs11Key.ToSessionObject(session, keyAttributes);
            }
            // получить атрибуты личного ключа
            Attribute[] privateKeyAttributes = PrivateKeyAttributes(
                null, privateKey, info
            ); 
            // проверить поддержку ключа
            if (privateKeyAttributes == null) throw new NotSupportedException(); 

            // указать тип ключа
            Attribute[] attributes = new Attribute[] {
	            CreateAttribute(API.CKA_CLASS, API.CKO_PRIVATE_KEY)
            };
            // добавить атрибуты личного ключа
            attributes = Attribute.Join(attributes, privateKeyAttributes); 

            // создать сеансовый объект
            return session.CreateObject(Attribute.Join(attributes, keyAttributes)); 
        }
		// атрибуты открытого ключа
		public abstract Attribute[] PublicKeyAttributes(
            Applet applet, IPublicKey publicKey, MechanismInfo info
        ); 
		// атрибуты личного ключа
		public abstract Attribute[] PrivateKeyAttributes(
            Applet applet, IPrivateKey privateKey, MechanismInfo info
        ); 
	    // атрибуты симметричного ключа
	    public virtual Attribute[] SecretKeyAttributes(SecretKeyFactory keyType, int keySize, bool hasValue) 
        { 
            // атрибуты созданного ключа
            if (hasValue) return new Attribute[] {
                  CreateAttribute(API.CKA_KEY_TYPE, API.CKK_GENERIC_SECRET) 
            }; 
            // атрибуты создаваемого ключа
            return new Attribute[] {
                CreateAttribute(API.CKA_KEY_TYPE , API.CKK_GENERIC_SECRET), 
                CreateAttribute(API.CKA_VALUE_LEN, (uint)keySize         ) 
            }; 
        }
		// создать ключ по сеансовому объекту
		public Attributes GetKeyAttributes(SessionObject obj, params Attribute[] attributes)
        {
            if (obj.OnToken())
            {
                // выделить память для атрибутов
                Attribute[] keyAttributes = new Attribute[] { 

	                // задать стандартные типы атрибутов
	                CreateAttribute(API.CKA_TOKEN,       API.CK_TRUE ), 
	                CreateAttribute(API.CKA_EXTRACTABLE, API.CK_TRUE ), 
	                CreateAttribute(API.CKA_SENSITIVE,   API.CK_FALSE), 
            
	                // задать стандартные типы атрибутов
                    new Attribute(API.CKA_CLASS   ), 
	                new Attribute(API.CKA_KEY_TYPE), 
                    new Attribute(API.CKA_ID      )
                }; 
                // указать дополнительные атрибуты
                attributes = Attribute.Join(keyAttributes, attributes); 
            }
            else {
                // выделить память для атрибутов
                Attribute[] keyAttributes = new Attribute[] { 

	                // задать стандартные типы атрибутов
	                CreateAttribute(API.CKA_TOKEN,       API.CK_FALSE), 
	                CreateAttribute(API.CKA_EXTRACTABLE, API.CK_TRUE ), 
	                CreateAttribute(API.CKA_SENSITIVE,   API.CK_FALSE), 
            
	                // задать стандартные типы атрибутов
                    new Attribute(API.CKA_CLASS   ), 
	                new Attribute(API.CKA_KEY_TYPE) 
                }; 
                // указать дополнительные атрибуты
                attributes = Attribute.Join(keyAttributes, attributes); 
            }
            // получить атрибуты объекта
            return new Attributes(obj.GetAttributes(attributes)); 
        }
    }
}

File: Source/Aladdin.CAPI.PKCS11.Java/src/aladdin/capi/pkcs11/Provider.java
package aladdin.capi.pkcs11;
import aladdin.*; 
import aladdin.capi.*; 
import aladdin.pkcs11.*; 
import java.io.*;
import java.util.*;

///////////////////////////////////////////////////////////////////////////
// Криптографический провайдер
///////////////////////////////////////////////////////////////////////////
public abstract class Provider extends aladdin.capi.CryptoProvider
{
    // конструктор
    public Provider(String name) { this.name = name; } 
    
    // имя провайдера
	@Override public final String name() { return name; } private final String name; 
    
	// интерфейс вызова функций
	public abstract Module module();  
    
    // возможность импорта ключевой пары в память
    public boolean canImportSessionPair(Applet applet) { return true; }

	///////////////////////////////////////////////////////////////////////
	// Управление устройствами
	///////////////////////////////////////////////////////////////////////
    @Override
	public String[] enumerateStores(Scope scope)
	{ 
        // проверить область видимости
        if (scope != Scope.SYSTEM) return new String[0]; 

        // выделить память для имен смарт-карт
        List<String> stores = new ArrayList<String>(); 
        try { 
            // получить список считывателей
            long[] slotList = module().getSlotList(true); 

            // для всех найденных смарт-карт
            for (int i = 0; i < slotList.length; i++) 
            {
                // создать объект считывателя
                try (Slot slot = new Slot(this, slotList[i])) 
                {
                    // проверить наличие смарт-карты
                    if (slot.getState() != aladdin.pcsc.ReaderState.CARD) continue; 
                    
                    // добавить смарт-карту в список
                    if (!stores.contains(slot.name())) stores.add(slot.name()); 
                }
                // обработать возможную ошибку
                catch (Throwable e) {} 
            }
        }
        // вернуть список смарт-карт
        catch (Throwable e) {} return stores.toArray(new String[stores.size()]); 
    } 
	@Override
	public final SecurityStore openStore(Scope scope, String storeName) throws IOException 
    { 
        // проверить область видимости
        if (scope != Scope.SYSTEM) throw new NoSuchElementException(); 

        // получить список считывателей
        long[] slotList = module().getSlotList(true); 

        // для всех найденных смарт-карт
        for (int i = 0; i < slotList.length; i++) 
        {
            // получить имя считывателя
            SlotInfo info = module().getSlotInfo(slotList[i]); 

            // проверить совпадение имен
            if (!info.slotDescription().equals(storeName)) continue; 
            
            // открыть объект смарт-карты
            return new Token(this, slotList[i]); 
        }
        // при ошибке выбросить исключение
        throw new NoSuchElementException(); 
    }
    // фабрика генераторов случайных данных
	@Override public IRandFactory createRandFactory(SecurityObject scope, boolean strong)
    {
	    // проверить область видимости
	    if (scope instanceof Container) return RefObject.addRef(((Container)scope).store());
	        
	    // проверить область видимости
	    if (scope instanceof Applet) return RefObject.addRef((Applet)scope); 
        
        // вернуть фабрика генераторов
        return RefObject.addRef(this);
    }
    // создать генератор случайных данных
    @Override public IRand createRand(Object window) throws IOException
    { 
        // получить список считывателей
        long[] slotList = module().getSlotList(true); 
	
        // для каждого считывателя
        for (int i = 0; i < slotList.length; i++)
        try {
            // получить информацию устройства
            SlotInfo slotInfo = module().getSlotInfo(slotList[i]); 
            
            // проверить наличие смарт-карты
            if ((slotInfo.flags() & API.CKF_TOKEN_PRESENT) == 0) continue;
            
            // получить информацию апплета
            TokenInfo tokenInfo = module().getTokenInfo(slotList[i]); 

            // проверить наличие генератора случайных данных
            if ((tokenInfo.flags() & API.CKF_RNG) == 0) continue; 
            
            // создать объект смарт-карты
            try (Token token = new Token(this, slotList[i]))
            {
                // создать объект апплета
                try (Applet applet = new Applet(token, slotList[i])) 
                { 
                    // вернуть генератор случайных данных
                    return applet.createRand(window); 
                }
            }
        }
        // вызвать базовую функцию
        catch (Throwable e) {} return super.createRand(window);
	}
	// найти устройство с реализацией алгоритма
	public final Applet findApplet(SecurityObject scope, 
        long algID, long usage, int keySize) throws IOException
	{
        // проверить область видимости
        if (scope instanceof Container) 
        {
            // получить апплет контейнера
            Applet applet = ((Container)scope).store(); 
            
            // проверить поддержку алгоритма
            if (!applet.supported(algID, usage, keySize)) return null; 
            
            // вернуть устройство для контейнера
	        return RefObject.addRef(applet);
        }
        // проверить область видимости
        if (scope instanceof Applet) { Applet applet = (Applet)scope;
        
            // проверить поддержку алгоритма
            if (!applet.supported(algID, usage, keySize)) return null; 
            
            // вернуть устройство 
	        return RefObject.addRef(applet);
        }
        // получить список считывателей
        long[] slotList = module().getSlotList(true); 

        // для каждого считывателя
        for (int i = 0; i < slotList.length; i++)
        try {
            // получить информацию устройства
            SlotInfo slotInfo = module().getSlotInfo(slotList[i]); 
            
            // проверить наличие смарт-карты
            if ((slotInfo.flags() & API.CKF_TOKEN_PRESENT) == 0) continue;
            
            // создать объект смарт-карты
            try (Token token = new Token(this, slotList[i]))
            {
                // создать объект апплета
                try (Applet applet = new Applet(token, slotList[i]))
                {
                    // проверить поддержку алгоритма
                    if (!applet.supported(algID, usage, keySize)) continue; 
                        
                    // вернуть найденный апплет
                    return RefObject.addRef(applet);
                }
            }
        }
        catch (Throwable e) {} return null;
    }
	///////////////////////////////////////////////////////////////////////
	// Особенности провайдера
	///////////////////////////////////////////////////////////////////////
	public byte[] generateSeed(Applet applet) throws IOException { return null; } 
	
	// преобразование типа ключей
	public SecretKey convertSecretKey(
        SessionObject object, SecretKeyFactory keyFactory) throws IOException
    {
        // получить атрибуты ключа
	    Attributes attributes = getKeyAttributes(object, 
            new Attribute(API.CKA_KEY_TYPE, API.CKK_GENERIC_SECRET)
        ); 
        // при возможности извлечения значения
        if ((Byte)attributes.get(API.CKA_EXTRACTABLE).value() != API.CK_FALSE && 
            (Byte)attributes.get(API.CKA_SENSITIVE  ).value() == API.CK_FALSE)
        {
            // получить значение ключа
            Attribute attribute = new Attribute(API.CKA_VALUE, object.getValue()); 
            
            // добавить атрибут в список
            attributes = attributes.join(new Attributes(attribute)); 
        }
        // при отсутствии на смарт-карте
        if ((Byte)attributes.get(API.CKA_TOKEN).value() == API.CK_FALSE)
        {
            // проверить наличие значения
            if (attributes.get(API.CKA_VALUE) == null)
            {
                // при ошибке выбросить исключение
                throw new aladdin.pkcs11.Exception(API.CKR_KEY_UNEXTRACTABLE); 
            }
            // создать ключ по сеансовому объекту
            return new SecretKey(null, keyFactory, attributes); 
        }
        else {
            // определить идентификатор слота
            long slotID = object.session().slotID(); 
            
            // создать объект смарт-карты
            try (Token token = new Token(this, slotID))
            {
                // указать апплет
                try (Applet applet = new Applet(token, slotID)) 
                {
                    // создать ключ по сеансовому объекту
                    return new SecretKey(applet, keyFactory, attributes); 
                }
            }
        }
    }
    // преобразовать тип открытого ключа
	public abstract IPublicKey convertPublicKey(
        Applet applet, SessionObject object) throws IOException;
    
    // преобразовать тип личного ключа
	public abstract PrivateKey convertPrivateKey(SecurityObject scope, 
        SessionObject object, IPublicKey publicKey) throws IOException;
    
	// преобразование типа ключей
	public final SessionObject toSessionObject(
        Session session, ISecretKey key, Attribute[] keyAttributes) throws IOException
    {
        // проверить тип ключа
        if (key instanceof SecretKey) return ((SecretKey)key).toSessionObject(session, keyAttributes); 
        
        // получить значение ключа
        if (key.value() == null) throw new aladdin.pkcs11.Exception(API.CKR_KEY_UNEXTRACTABLE); 
        
        // создать список атрибутов
        Attribute[] attributes = new Attribute[] {

            // указать тип ключа
            new Attribute(API.CKA_CLASS, API.CKO_SECRET_KEY), 

            // указать тип ключа
            new Attribute(API.CKA_KEY_TYPE, API.CKK_GENERIC_SECRET), 

            // указать извлекаемость значения
            new Attribute(API.CKA_EXTRACTABLE, API.CK_TRUE ), 
            new Attribute(API.CKA_SENSITIVE  , API.CK_FALSE), 

            // указать значение ключа
            new Attribute(API.CKA_VALUE, key.value()) 
        }; 
        // добавить атрибуты ключа
        attributes = Attribute.join(attributes, keyAttributes); 
        
        // создать сеансовый объект
        return session.createObject(attributes, null); 
    }
	public final SessionObject toSessionObject(Session session, 
        IPublicKey publicKey, MechanismInfo info, Attribute[] keyAttributes) throws IOException
    {
        // получить атрибуты открытого ключа
        Attribute[] publicKeyAttributes = publicKeyAttributes(null, publicKey, info); 

        // проверить поддержку ключа
        if (publicKeyAttributes == null) throw new UnsupportedOperationException(); 

        // указать тип ключа
        Attribute[] attributes = new Attribute[] {
            new Attribute(API.CKA_CLASS, API.CKO_PUBLIC_KEY)
        }; 
        // добавить атрибуты открытого ключа
        attributes = Attribute.join(attributes, publicKeyAttributes); 
        
        // добавить атрибуты открытого ключа
        attributes = Attribute.join(attributes, keyAttributes); 
        
        // создать сеансовый объект
        return session.createObject(attributes, null); 
    }
	public final SessionObject toSessionObject(Session session, 
        IPrivateKey privateKey, MechanismInfo info, Attribute[] keyAttributes) throws IOException
    {
        // проверить тип ключа
        if (privateKey instanceof PrivateKey)
        {
            // создать сеансовый объект
            return ((PrivateKey)privateKey).toSessionObject(session, keyAttributes); 
        }
        // получить атрибуты личного ключа
        Attribute[] privateKeyAttributes = privateKeyAttributes(null, privateKey, info); 

        // проверить поддержку ключа
        if (privateKeyAttributes == null) throw new UnsupportedOperationException(); 

        // указать тип ключа
        Attribute[] attributes = new Attribute[] {
            new Attribute(API.CKA_CLASS, API.CKO_PRIVATE_KEY)
        }; 
        // добавить атрибуты личного ключа
        attributes = Attribute.join(attributes, privateKeyAttributes); 
        
        // добавить атрибуты личного ключа
        attributes = Attribute.join(attributes, keyAttributes); 
        
        // создать сеансовый объект
        return session.createObject(attributes, null); 
    }
	// атрибуты открытого и личного ключа
	public abstract Attribute[] publicKeyAttributes(
        Applet applet, IPublicKey  publicKey,  MechanismInfo info) throws IOException;  
	public abstract Attribute[] privateKeyAttributes(
        Applet applet, IPrivateKey privateKey, MechanismInfo info) throws IOException; 
    
	// атрибуты симметричного ключа
	public Attribute[] secretKeyAttributes(
        SecretKeyFactory keyFactory, int keySize, boolean hasValue) 
    { 
        // атрибуты созданного ключа
        if (hasValue) return new Attribute[] {
            new Attribute(API.CKA_KEY_TYPE , API.CKK_GENERIC_SECRET) 
        }; 
        // атрибуты создаваемого ключа
        return new Attribute[] {
            new Attribute(API.CKA_KEY_TYPE , API.CKK_GENERIC_SECRET), 
            new Attribute(API.CKA_VALUE_LEN, keySize               ) 
        }; 
    }
    // получить атрибуты ключа
    public Attributes getKeyAttributes(SessionObject object, Attribute... attributes) throws IOException
    {
        if (object.onToken())
        {
            // выделить память для атрибутов
            Attribute[] keyAttributes = new Attribute[] { 

                // задать стандартные типы атрибутов
                new Attribute(API.CKA_TOKEN      , API.CK_TRUE ), 
                new Attribute(API.CKA_EXTRACTABLE, API.CK_TRUE ),  
                new Attribute(API.CKA_SENSITIVE  , API.CK_FALSE),  

                // задать стандартные типы атрибутов
                new Attribute(API.CKA_CLASS   , Long  .class), 
                new Attribute(API.CKA_KEY_TYPE, Long  .class), 
                new Attribute(API.CKA_ID      , byte[].class)
            }; 
            // указать дополнительные атрибуты
            attributes = Attribute.join(keyAttributes, attributes);
        }
        else {
            // выделить память для атрибутов
            Attribute[] keyAttributes = new Attribute[] { 

                // задать стандартные типы атрибутов
                new Attribute(API.CKA_TOKEN      , API.CK_FALSE), 
                new Attribute(API.CKA_EXTRACTABLE, API.CK_TRUE ),  
                new Attribute(API.CKA_SENSITIVE  , API.CK_FALSE),  

                // задать стандартные типы атрибутов
                new Attribute(API.CKA_CLASS   , Long  .class), 
                new Attribute(API.CKA_KEY_TYPE, Long  .class), 
            }; 
            // указать дополнительные атрибуты
            attributes = Attribute.join(keyAttributes, attributes);
        }
        // получить атрибуты объекта
	    return new Attributes(object.getAttributes(attributes)); 
    }
} 
